libraries/standard/mqtt/src/mqtt.c

```diff
@@ -439,56 +445,87 @@ static bool matchWildcards( const char * pTopicName,
+            /* Exit the loop if we hit the level separator. */
+            if( pTopicName[ *pNameIndex ] == '/' )
             {
-                ( *pNameIndex )++;
+                nextLevelExistsInTopicName = true;
+                break;
             }
 
-            /* Decrement the topic name index for 2 different cases:
-             * - If the break condition is ( *pNameIndex < topicNameLength ), then
-             *   we have reached past the end of the topic name and we move back the
-             *   the index on the last character.
-             * - If the break condition is ( pTopicName[ *pNameIndex ] != '/' ), we
-             *   move back the index on the '/' character. */
-            ( *pNameIndex )--;
+            ( *pNameIndex )++;
         }
 
-        /* '#' matches everything remaining in the topic name. It must be the
-         * last character in a topic filter. */
-        else if( ( pTopicFilter[ filterIndex ] == '#' ) &&
-                 ( filterIndex == ( topicFilterLength - 1U ) ) )
+        /* Determine if the topic filter contains a child level after the current level
+         * represented by the '+' wildcard. */
+        if( ( *pFilterIndex < ( topicFilterLength - 1U ) ) &&
+            ( pTopicFilter[ *pFilterIndex + 1U ] == '/' ) )
         {
-            /* Subsequent characters don't need to be checked for the
-             * multi-level wildcard. */
-            *pMatch = true;
-            shouldStopMatching = true;
+            nextLevelExistsinTopicFilter = true;
         }
-        else
+
+        /* If the topic name contains a child level but the topic filter ends at
+         * the current level, then there does not exist a match. */
+        if( ( nextLevelExistsInTopicName == true ) &&
+            ( nextLevelExistsinTopicFilter == false ) )
         {
-            /* Any character mismatch other than '+' or '#' means the topic
-             * name does not match the topic filter. */
             *pMatch = false;
             shouldStopMatching = true;
         }
+
+        /* If the topic name and topic filter have child levels, then advance the
+         * filter index to the level separator in the topic filter, so that match
+         * can be performed in the next level.
+         * Note: The name index already points to the level separator in the topic
+         * name. */
+        else if( ( nextLevelExistsInTopicName == true ) &&
+                 ( nextLevelExistsinTopicFilter == true ) )
+        {
+            ( *pFilterIndex )++;
+        }
+        else
+        {
+            /* If we have reached here, the the loop terminated on the
+             * ( *pNameIndex < topicNameLength) condition, which means that have
+             * reached past the end of the topic name, and thus, we decrement the
+             * index to the last character in the topic name.*/
+            ( *pNameIndex )--;
```

Comment:
The decrement of `*pNameIndex` has been reduced to only the case when it has reached the end of the topic name. (In the earlier revision, this was decremented also for the case when the topic name contained a child level beyond the current level. The decrement for a special scenario in that case led to underflow)

Reply:
Are the changes to this file necessary? 

`pNameIndex` is a pointer to a `uint16_t`, so per [6.2.5.9](http://www.open-std.org/JTC1/sc22/wg14/www/docs/n1256.pdf) of the C standard it can't under/overflow; computation regarding unsigned types is well defined, and for a uint16_t, is reduced modulo UINT16_MAX+1. While this decrement would turn a 0 to UINT16_MAX, we increment `nameIndex` back to 0 before the end of the `while` loop in `matchTopicFilter`. 

So, it seems to me like this case is already handled, and extra code doesn't seem necessary if a comment would be sufficient.

Reply:
> pNameIndex is a pointer to a uint16_t, so per 6.2.5.9 of the C standard it can't under/overflow; computation regarding unsigned types is well defined, and for a uint16_t, is reduced modulo UINT16_MAX+1. While this decrement would turn a 0 to UINT16_MAX, we increment nameIndex back to 0 before the end of the while loop in matchTopicFilter.

CBMC flags underflow error for decrementing 0 to 65535 in `matchWildcards` and increment 65535 to 0 in `matchTopicFilter` as overflow which is a completely fair expectation of a function to not have such side-effects on its values. The functions are independently held accountable for their quality, it just so happens that the underflow in `matchWildcards` happens to be countered by the overflow in `matchTopicFilter`, and thus, the over logic works fine. However, we shouldn't depend on such a behavior for correct functionality of the topic matching algorithm

Reply:
If this is a CBMC issue then I understand why it was changed; I had commented because I don't think we need to over engineer what already works as expected, since doing so can make the code harder to read, and can introduce mistakes, e.g. for MISRA.